File: services/multiagent/workflow.ts
import type { Worksheet, SavedWorksheet } from '../../types';
import { AdpAgent } from './agents/adpAgent';
import { AcAgent } from './agents/acAgent';
import { EvaluatorAgent } from './agents/evaluatorAgent';
import { searchPictograms } from '../pictogramService';
import { normalizeWorksheet, getFlattenedItemsFromExercise } from '../worksheetNormalizer';
import { getActiveProfileData, getAppData, parseJsonPayload, repairJsonResponse } from './utils';
import { runAiPrompt } from '../aiClient';
import { buildGenerationPromptBase } from './prompts/shared';
import {
  getAvailableExerciseTypes,
  getExerciseSchema,
} from '../../components/exercises/registry';

export interface GenerateWorksheetOptions {
  topic?: string;
  goal?: string;
  extraDetails?: string;
  language?: 'es' | 'val' | 'en';
}

const EXERCISE_COUNT_PATTERNS = [
  /\b(\d{1,2})\s+(?:ejercicios?|actividades?|secciones?)\b/i,
  /\b(?:con|de|tenga|tener|incluye?|incluya|quiero|necesito)\s+(\d{1,2})\s+(?:ejercicios?|actividades?|secciones?)\b/i,
  /\b(\d{1,2})\s+(?:exercises?|activities?|sections?)\b/i,
];

const getSemanticContext = (options: GenerateWorksheetOptions): string =>
  [options.topic, options.goal, options.extraDetails]
    .filter(Boolean)
    .join(' ')
    .toLowerCase();

const extractRequestedExerciseCount = (options: GenerateWorksheetOptions): number | undefined => {
  const textBlocks = [
    options.topic,
    options.goal,
    options.extraDetails,
  ].filter(Boolean) as string[];

  for (const text of textBlocks) {
    for (const pattern of EXERCISE_COUNT_PATTERNS) {
      const match = text.match(pattern);
      if (!match) continue;

      const count = Number.parseInt(match[1], 10);
      if (Number.isFinite(count) && count > 0) {
        return count;
      }
    }
  }

  return undefined;
};

/**
 * Orchestrates the Multi-Agent System (MAS) to design, construct and assemble a worksheet.
 */
export const generateWorksheet = async (options: GenerateWorksheetOptions): Promise<Worksheet> => {
  const { content: childProfile, showPictogramInstructions } = getActiveProfileData();
  const startTime = Date.now();
  
  const requestedLanguage = options.language || 'es';
  const availableExerciseTypes = getAvailableExerciseTypes();
  const requestedExerciseCount = extractRequestedExerciseCount(options);

  const appData = getAppData();
  const useSinglePrompt = appData?.aiSettings.useSinglePrompt || false;

  let adpTimeMs = 0;
  let acTimeMs = 0;
  let retryCount = 0;

  try {
    if (useSinglePrompt) {
      const promptText = buildGenerationPromptBase(
        "Generate a complete worksheet visual adaptada.",
        {
          topic: options.topic,
          goal: options.goal,
          extraDetails: options.extraDetails,
          requestedExerciseCount,
          language: requestedLanguage,
        },
        childProfile,
        showPictogramInstructions
      );

      let rawText = '';
      let parsedPayload: any = null;
      try {
        rawText = await runAiPrompt(promptText);
        parsedPayload = parseJsonPayload(rawText);
      } catch (err) {
        console.warn("Fallo en generación single-prompt, intentando reparar JSON:", err);
        try {
          parsedPayload = await repairJsonResponse(rawText, 'worksheet');
        } catch (repairErr) {
          throw new Error("No se pudo generar la ficha en una única llamada de IA debido a un fallo estructurado irrecuperable.");
        }
      }

      // Normalizar la ficha
      let worksheet = normalizeWorksheet(parsedPayload as Worksheet);
      worksheet.originalTopic = options.topic;
      worksheet.originalGoal = options.goal;
      worksheet.originalExtraDetails = options.extraDetails;
      worksheet.language = requestedLanguage;

      // Resolver pictogramas
      const searchTerms: { type: 'main' | 'section_instruction' | 'item'; path: number[]; term: string }[] = [];
      if (worksheet.pictogramSearchTerm) {
        searchTerms.push({ type: 'main', path: [], term: worksheet.pictogramSearchTerm });
      }

      worksheet.sections.forEach((section, sectionIndex) => {
        (section.instruction?.pictograms || []).forEach((picto, pictoIndex) => {
          searchTerms.push({
            type: 'section_instruction',
            path: [sectionIndex, pictoIndex],
            term: picto.searchTerm || picto.content,
          });
        });

        (section.items || []).forEach((item, itemIndex) => {
          if (item.type === 'image' || item.type === 'traceable_text') {
            searchTerms.push({
              type: 'item',
              path: [sectionIndex, itemIndex],
              term: item.searchTerm || item.content,
            });
          }
        });
      });

      const pictoSearchPromises = searchTerms.map(st => searchPictograms(st.term, requestedLanguage));
      const pictoSearchResults = await Promise.all(pictoSearchPromises);

      pictoSearchResults.forEach((foundPictos, idx) => {
        const st = searchTerms[idx];
        const urls = foundPictos.map(p => p.url);
        const chosenUrl = urls.length > 0 ? urls[0] : '';

        if (st.type === 'main') {
          worksheet.pictoOptions = urls;
          worksheet.selectedPictoUrl = chosenUrl;
        } else if (st.type === 'section_instruction') {
          const [sectionIndex, pictoIndex] = st.path;
          const pictoObj = worksheet.sections[sectionIndex].instruction.pictograms?.[pictoIndex];
          if (pictoObj) pictoObj.url = chosenUrl;
        } else if (st.type === 'item') {
          const [sectionIndex, itemIndex] = st.path;
          const itemObj = worksheet.sections[sectionIndex].items[itemIndex];
          if (itemObj) {
            itemObj.pictoOptions = urls;
            itemObj.selectedPictoUrl = chosenUrl;
          }
        }
      });

      worksheet = normalizeWorksheet(worksheet);
      worksheet.telemetry = {
        generationTimeMs: Date.now() - startTime,
        adpTimeMs: 0,
        acTimeMs: 0,
        rejectionCount: 0,
        manualEditsCount: 0,
        pictoOverridesCount: 0,
        retryCount: 0,
        createdTimestamp: new Date().toISOString(),
      };

      return worksheet;
    }
    // 1. Etapa 1: Agente Diseñador Pedagógico (ADP)
    const adpStartTime = Date.now();
    const adpAgent = new AdpAgent();
    const blueprint = await adpAgent.generateBlueprint(
      {
        topic: options.topic,
        goal: options.goal,
        extraDetails: options.extraDetails,
        requestedExerciseCount,
        language: requestedLanguage,
      },
      childProfile,
      availableExerciseTypes
    );
    adpTimeMs = Date.now() - adpStartTime;

    const exercisePlans = blueprint.exercisePlans || [];

    // 2. Etapa 2: Agente Constructor (AC) + Evaluador Individual (PEA) - Ejecución en paralelo con retry
    const acStartTime = Date.now();
    const acAgent = new AcAgent();
    const evaluatorAgent = new EvaluatorAgent();

    const sectionPromises = exercisePlans.map(async (plan, index) => {
      const typeKey = String(plan.type || 'rodear').toLowerCase().trim() as any;
      let targetSchema = '';
      try {
        targetSchema = getExerciseSchema(typeKey);
      } catch {
        targetSchema = getExerciseSchema('rodear');
      }

      let attempts = 0;
      const maxAttempts = 3;
      let generatedSection: any = null;
      let individualFeedback: string | undefined = undefined;

      while (attempts < maxAttempts) {
        try {
          generatedSection = await acAgent.generateExercise(
            plan,
            index,
            targetSchema,
            requestedLanguage,
            () => { retryCount++; },
            individualFeedback
          );

          // Evaluar individualmente el ejercicio generado
          const evalResult = await evaluatorAgent.evaluateExercise(
            plan,
            generatedSection,
            childProfile,
            requestedLanguage
          );

          if (evalResult.approved) {
            return generatedSection;
          } else {
            console.warn(
              `[Evaluador Individual] Ejercicio ${index + 1} de tipo ${typeKey} rechazado. Feedback: ${evalResult.feedback}`
            );
            individualFeedback = evalResult.feedback;
            attempts++;
            retryCount++;
          }
        } catch (error) {
          attempts++;
          retryCount++;
          console.warn(
            `Intento ${attempts} fallido para generar/evaluar el ejercicio ${index + 1} de tipo ${typeKey}. Error:`,
            error
          );
          if (attempts >= maxAttempts) {
            throw error;
          }
        }
      }
      throw new Error(`Fallo tras ${maxAttempts} intentos al generar el ejercicio de tipo "${typeKey}".`);
    });

    const rawSections = await Promise.all(sectionPromises);
    acTimeMs = Date.now() - acStartTime;

    // 3. Etapa 3: Ensamblador de Fichas (Código Puro)
    const assembledSections = rawSections.map((sectionPayload) => {
      return normalizeWorksheet({
        title: blueprint.title,
        pictogramSearchTerm: blueprint.pictogramSearchTerm,
        sections: [sectionPayload],
      }).sections[0];
    });

    // Crear el objeto provisional de la ficha
    let worksheet: Worksheet = {
      title: blueprint.title || options.topic || "Ficha adaptada",
      pictogramSearchTerm: blueprint.pictogramSearchTerm || options.topic || "ficha",
      sections: assembledSections,
      originalTopic: options.topic,
      originalGoal: options.goal,
      originalExtraDetails: options.extraDetails,
      language: requestedLanguage,
    };

    // 4. Etapa 4: Evaluador Global (GPEA)
    let globalAttempts = 0;
    const maxGlobalAttempts = 2;
    let globalApproved = false;

    while (globalAttempts < maxGlobalAttempts && !globalApproved) {
      const globalEval = await evaluatorAgent.evaluateWorksheet(
        childProfile,
        options.topic || '',
        options.goal || '',
        worksheet,
        requestedLanguage
      );

      if (globalEval.approved || !globalEval.exerciseModifications || globalEval.exerciseModifications.length === 0) {
        globalApproved = true;
      } else {
        console.warn(`[Evaluador Global] Ficha rechazada. Issues: ${globalEval.globalIssues}`);
        globalAttempts++;
        retryCount++;

        // Aplicar las correcciones solicitadas en paralelo
        const correctionPromises = globalEval.exerciseModifications.map(async (mod) => {
          const idx = mod.exerciseIndex;
          if (idx < 0 || idx >= worksheet.sections.length) return;

          const plan = exercisePlans[idx];
          const typeKey = String(plan.type || 'rodear').toLowerCase().trim() as any;
          let targetSchema = '';
          try {
            targetSchema = getExerciseSchema(typeKey);
          } catch {
            targetSchema = getExerciseSchema('rodear');
          }

          console.warn(`[Evaluador Global] Corrigiendo ejercicio ${idx + 1} de tipo ${typeKey} con feedback global.`);
          const correctedSection = await acAgent.generateExercise(
            plan,
            idx,
            targetSchema,
            requestedLanguage,
            () => { retryCount++; },
            mod.feedback
          );

          // Normalizar el ejercicio corregido
          worksheet.sections[idx] = normalizeWorksheet({
            title: blueprint.title,
            pictogramSearchTerm: blueprint.pictogramSearchTerm,
            sections: [correctedSection],
          }).sections[0];
        });

        await Promise.all(correctionPromises);
      }
    }

    // Resolver pictogramas
    const searchTerms: { type: 'main' | 'section_instruction' | 'item'; path: number[]; term: string }[] = [];

    // Pictograma de cabecera
    if (worksheet.pictogramSearchTerm) {
      searchTerms.push({
        type: 'main',
        path: [],
        term: worksheet.pictogramSearchTerm,
      });
    }

    // Pictogramas de enunciados y elementos de ejercicios
    worksheet.sections.forEach((section, sectionIndex) => {
      // Enunciado
      (section.instruction?.pictograms || []).forEach((picto, pictoIndex) => {
        searchTerms.push({
          type: 'section_instruction',
          path: [sectionIndex, pictoIndex],
          term: picto.searchTerm || picto.content,
        });
      });

      // Elementos de la sección
      (section.items || []).forEach((item, itemIndex) => {
        if (item.type === 'image' || item.type === 'traceable_text') {
          searchTerms.push({
            type: 'item',
            path: [sectionIndex, itemIndex],
            term: item.searchTerm || item.content,
          });
        }
      });
    });

    // Resolver en paralelo todas las búsquedas de pictogramas a ARASAAC
    const pictoSearchPromises = searchTerms.map(st => searchPictograms(st.term, requestedLanguage));
    const pictoSearchResults = await Promise.all(pictoSearchPromises);

    pictoSearchResults.forEach((foundPictos, idx) => {
      const st = searchTerms[idx];
      const urls = foundPictos.map(p => p.url);
      const chosenUrl = urls.length > 0 ? urls[0] : '';

      if (st.type === 'main') {
        worksheet.pictoOptions = urls;
        worksheet.selectedPictoUrl = chosenUrl;
      } else if (st.type === 'section_instruction') {
        const [sectionIndex, pictoIndex] = st.path;
        const pictoObj = worksheet.sections[sectionIndex].instruction.pictograms?.[pictoIndex];
        if (pictoObj) {
          pictoObj.url = chosenUrl;
        }
      } else if (st.type === 'item') {
        const [sectionIndex, itemIndex] = st.path;
        const itemObj = worksheet.sections[sectionIndex].items[itemIndex];
        if (itemObj) {
          itemObj.pictoOptions = urls;
          itemObj.selectedPictoUrl = chosenUrl;
        }
      }
    });

    // Normalizar la ficha entera con IDs internos
    worksheet = normalizeWorksheet(worksheet);

    const totalTimeMs = Date.now() - startTime;

    // Rellenar telemetría
    worksheet.telemetry = {
      generationTimeMs: totalTimeMs,
      adpTimeMs,
      acTimeMs,
      rejectionCount: 0,
      manualEditsCount: 0,
      pictoOverridesCount: 0,
      retryCount,
      createdTimestamp: new Date().toISOString(),
    };

    return worksheet;
  } catch (error) {
    console.error('Error al generar la ficha mediante MAS:', error);
    throw new Error(error instanceof Error ? error.message : 'No se pudo generar la ficha adaptada. Inténtalo de nuevo.');
  }
};

File: scripts/generator/workflow.py
import re
import json
import time

from .prompts import (
    build_adp_prompt,
    build_ac_prompt,
    build_pea_prompt,
    build_gpea_prompt,
    build_single_prompt
)
from .api import (
    call_llm,
    extract_json_object,
    repair_json_response,
    resolve_pictograms_in_worksheet
)

def get_semantic_exercise_count(options):
    text_blocks = [options.get("topic", ""), options.get("goal", ""), options.get("extraDetails", "")]
    patterns = [
        r"\b(\d{1,2})\s+(?:ejercicios?|actividades?|secciones?)\b",
        r"\b(?:con|de|tenga|tener|incluye?|incluya|quiero|necesito)\s+(\d{1,2})\s+(?:ejercicios?|actividades?|secciones?)\b",
        r"\b(\d{1,2})\s+(?:exercises?|activities?|sections?)\b"
    ]
    for text in text_blocks:
        if not text:
            continue
        for pat in patterns:
            match = re.search(pat, text, re.IGNORECASE)
            if match:
                val = int(match.group(1))
                if val > 0:
                    return val
    return None

def normalize_exercise_structure(section):
    """
    Normalizes sections and copies structural prompts/copies lists to items array to match TS.
    """
    exercise_type = section.get("exerciseType")
    exercise = section.get("exercise", {})
    
    items = []
    if exercise_type == "repasar":
        prompts = exercise.get("prompts", [])
        for prompt in prompts:
            items.append({
                "type": "traceable_text",
                "content": str(prompt).upper()
            })
            
    elif exercise_type == "copiar":
        copies = exercise.get("copies", [])
        for copy in copies:
            items.append({
                "type": "traceable_text",
                "content": str(copy).upper()
            })
            
    elif exercise_type == "rodear":
        options = exercise.get("options", [])
        for opt in options:
            items.append({
                "type": "image",
                "content": opt.get("content", ""),
                "searchTerm": opt.get("searchTerm", opt.get("content", "")).lower(),
                "selectedPictoUrl": "",
                "pictoOptions": [],
                "pictogramRenderMode": "auto"
            })
            
    elif exercise_type == "unir":
        pairs = exercise.get("pairs", [])
        left_items = []
        right_items = []
        for pair in pairs:
            left = pair.get("left", {})
            right = pair.get("right", {})
            left_items.append({
                "type": "image",
                "content": left.get("content", ""),
                "searchTerm": left.get("searchTerm", left.get("content", "")).lower(),
                "selectedPictoUrl": "",
                "pictoOptions": [],
                "pictogramRenderMode": "auto"
            })
            right_items.append({
                "type": "image",
                "content": right.get("content", ""),
                "searchTerm": right.get("searchTerm", right.get("content", "")).lower(),
                "selectedPictoUrl": "",
                "pictoOptions": [],
                "pictogramRenderMode": "auto"
            })
        items = left_items + right_items
        
    section["items"] = items

def generate_worksheet(options, profile_content, show_pictograms, settings):
    """
    Orchestrates the Worksheet generation (Single-Prompt or Multi-Agent System).
    """
    start_time = time.time()
    language = options.get("language", "es")
    arasaac_url = settings.get("arasaacApiUrl", "https://api.arasaac.org/api/pictograms")
    
    use_single_prompt = settings.get("useSinglePrompt", False)
    
    adp_time_ms = 0
    ac_time_ms = 0
    retry_count = 0
    
    if use_single_prompt:
        # SINGLE PROMPT MODE
        prompt = build_single_prompt(options, profile_content, language, show_pictograms)
        raw_output = call_llm(prompt, settings)
        
        try:
            worksheet = json.loads(extract_json_object(raw_output))
        except Exception:
            # Try to repair JSON
            worksheet = repair_json_response(raw_output, "Worksheet JSON containing sections list", settings)
            
        # Structure sections items
        for sec in worksheet.get("sections", []):
            normalize_exercise_structure(sec)
    else:
        # MULTI-AGENT SYSTEM (MAS) MODE
        options["requestedExerciseCount"] = get_semantic_exercise_count(options)
        
        # 1. ADP (Planning)
        adp_start = time.time()
        adp_prompt = build_adp_prompt(options, profile_content, language)
        adp_raw = call_llm(adp_prompt, settings)
        
        try:
            blueprint = json.loads(extract_json_object(adp_raw))
        except Exception:
            blueprint = repair_json_response(adp_raw, "ADP Worksheet Blueprint JSON containing exercisePlans list", settings)
        adp_time_ms = int((time.time() - adp_start) * 1000)
        
        exercise_plans = blueprint.get("exercisePlans", [])
        sections = []
        
        # 2. AC + PEA (Construction & Individual Evaluation)
        ac_start = time.time()
        for idx, plan in enumerate(exercise_plans):
            feedback = None
            section_json = None
            
            # Max 3 attempts per exercise section
            for attempt in range(3):
                ac_prompt = build_ac_prompt(plan, language, feedback)
                ac_raw = call_llm(ac_prompt, settings)
                
                try:
                    section_json = json.loads(extract_json_object(ac_raw))
                except Exception:
                    section_json = repair_json_response(ac_raw, f"Exercise JSON for type '{plan['type']}'", settings)
                
                # Run PEA audit
                pea_prompt = build_pea_prompt(plan, json.dumps(section_json), profile_content, language)
                pea_raw = call_llm(pea_prompt, settings)
                
                try:
                    audit_res = json.loads(extract_json_object(pea_raw))
                except Exception:
                    audit_res = {"approved": True, "feedback": ""} # fallback
                    
                if audit_res.get("approved", True):
                    break
                else:
                    feedback = audit_res.get("feedback", "El ejercicio no cumple las pautas.")
                    retry_count += 1
            
            if section_json:
                normalize_exercise_structure(section_json)
                sections.append(section_json)
                
        ac_time_ms = int((time.time() - ac_start) * 1000)
        
        # Assemble preliminary worksheet
        worksheet = {
            "title": blueprint.get("title", "Ficha Adaptada"),
            "pictogramSearchTerm": blueprint.get("pictogramSearchTerm", ""),
            "sections": sections
        }
        
        # 3. GPEA (Global Evaluation)
        # Max 2 global attempts
        for global_attempt in range(2):
            gpea_prompt = build_gpea_prompt(profile_content, json.dumps(worksheet), language)
            gpea_raw = call_llm(gpea_prompt, settings)
            
            try:
                gpea_audit = json.loads(extract_json_object(gpea_raw))
            except Exception:
                gpea_audit = {"approved": True, "rejectedSectionIndexes": []}
                
            if gpea_audit.get("approved", True) or not gpea_audit.get("rejectedSectionIndexes"):
                break
                
            rejected_indexes = gpea_audit.get("rejectedSectionIndexes", [])
            feedback = gpea_audit.get("feedback", "Coherencia global inválida.")
            
            # Reconstruct failed sections
            for rej_idx in rejected_indexes:
                if rej_idx < len(exercise_plans) and rej_idx < len(worksheet["sections"]):
                    plan = exercise_plans[rej_idx]
                    ac_prompt = build_ac_prompt(plan, language, f"GLOBAL REJECTION FEEDBACK: {feedback}")
                    ac_raw = call_llm(ac_prompt, settings)
                    
                    try:
                        section_json = json.loads(extract_json_object(ac_raw))
                    except Exception:
                        section_json = repair_json_response(ac_raw, f"Exercise JSON for type '{plan['type']}'", settings)
                        
                    normalize_exercise_structure(section_json)
                    worksheet["sections"][rej_idx] = section_json
                    retry_count += 1
                    
    # 4. Resolve Pictograms (ARASAAC lookup in Python)
    resolve_pictograms_in_worksheet(worksheet, language, arasaac_url)
    
    total_time_ms = int((time.time() - start_time) * 1000)
    
    # Embed telemetry
    worksheet["telemetry"] = {
        "generationTimeMs": total_time_ms,
        "adpTimeMs": adp_time_ms,
        "acTimeMs": ac_time_ms,
        "rejectionCount": 0,
        "manualEditsCount": 0,
        "pictoOverridesCount": 0,
        "retryCount": retry_count,
        "createdTimestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    }
    
    return worksheet
